core/src/main/java/org/kohsuke/stapler/EvaluationTrace.java

```diff
@@ -61,4 +67,38 @@ public static EvaluationTrace get(StaplerRequest req) {
      * Used for counting trace header.
      */
     private static final String KEY = EvaluationTrace.class.getName();
+
+    public abstract static class ApplicationTracer {
+        protected abstract void record(StaplerRequest req, String message);
+
+        public static void trace(StaplerRequest req, String message) {
+            List<ApplicationTracer> tracers = getTracers();
+            for (ApplicationTracer tracer : tracers) {
+                tracer.record(req, message);
+            }
+        }
+
+        private static volatile List<ApplicationTracer> tracers;
+
+        @Nonnull
+        private static List<ApplicationTracer> getTracers() {
+            synchronized (ApplicationTracer.class) {
+                if (tracers == null) {
+                    List<ApplicationTracer> t = new ArrayList<>();
+                    for (ApplicationTracer tracer : ServiceLoader.load(EvaluationTrace.ApplicationTracer.class, Stapler.getCurrent().getWebApp().getClassLoader())) {
+                        try {
+                            t.add(tracer);
+                        } catch (Exception e) {
+                            // robustness
+                            if (LOGGER.isLoggable(Level.FINE)) {
+                                LOGGER.log(Level.FINE, "Exception thrown when notifying tracer", e);
+                            }
+                        }
+                    }
+                    tracers = t;
+                }
+            }
+            return tracers;
```

Comment:
This should really be _inside_ the `synchronized` block. I _think_ it is harmless to have it outside, since there will be a synchronization barrier before, but definitely not the best style.